src/middleware/jwtAuthentication.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+import jwt from 'jsonwebtoken'
+import logger from 'winston'
+
+import * as client from '../model/clients'
+import * as configIndex from '../config'
+import fs from 'fs'
+import path from 'path'
+
+const TOKEN_PATTERN = /^ *(?:[Bb][Ee][Aa][Rr][Ee][Rr]) +([A-Za-z0-9\-._~+/]+=*) *$/
+
+async function authenticateClient(clientID) {
+  return client.ClientModel.findOne({ clientID }).then((client) => {
+    if (!client) {
+      throw new Error('Client does not exist')
+    }
+    return client
+  })
+}
+
+function resolveJwtSecretOrPublicKey() {
+  let secretOrPublicKey = configIndex.config.get(
+    'authentication:jwt:secretOrPublicKey'
+  )
+
+  try {
+    const publicKeyFilePath = path.resolve(
+      __dirname,
+      '..',
+      '..',
+      'resources',
+      'certs',
+      'jwt',
+      secretOrPublicKey
+    )
+
+    // Check file exists
+    if (fs.existsSync(publicKeyFilePath)) {
+      secretOrPublicKey = fs.readFileSync(publicKeyFilePath).toString()
+    }
+    return secretOrPublicKey
+  } catch (error) {
+    throw new Error(
+      `Could not read public key file to verify asymmetric JWT: ${error}`
+    )
+  }
+}
+
+function getJwtOptions() {
+  const jwtConfig = configIndex.config.get('authentication:jwt')
+  const jwtOptions = {}
+
+  // Algorithms can be input as an environment variable therefore the string needs to be split
+  if (jwtConfig.algorithms) {
+    jwtOptions.algorithms = jwtConfig.algorithms.split(' ')
+  } else {
+    // The jsonwebtoken package does not require this field, but allowing any algorithm to be used opens a security risk
+    throw new Error('JWT Algorithm not specified')
+  }
+
+  // Audience can be input as an environment variable therefore the string needs to be split
+  if (jwtConfig.audience) {
+    jwtOptions.audience = jwtConfig.audience.split(' ')
+  }
+
+  jwtOptions.issuer = jwtConfig.issuer
+
+  return jwtOptions
+}
+
+async function authenticateToken(ctx) {
+  if (ctx.authenticated != null) {
+    return
+  }
+
+  const authHeader = ctx.request.header.authorization || ''
+  const token = TOKEN_PATTERN.exec(authHeader)
+
+  if (!token) {
+    logger.warn(`Missing or invalid 'Authorization' header`)
+    return
+  }
+
+  try {
+    const decodedToken = jwt.verify(
+      token[1],
+      resolveJwtSecretOrPublicKey(),
+      getJwtOptions()
+    )
+
+    if (!decodedToken.sub) {
+      logger.error('Invalid JWT Payload')
```

Comment:
Maybe we need to highlight a bit more why this error happened. We can state `No JWT sub supplied in payload`

Reply:
@BMartinos Is it safe to log these details? I was being cautious with how much I logged about the auth issue. Heath appears to deliberately obscure the auth error messages.
Would an attacker ever have access to the logs but not the server instance? Probably not I guess...

Reply:
The only way they would have access to the logs if they have access to the core server to print the logs, or a user within the console where they can view the logs. The main reason for adding in the extra bit is to notify the implementer of a missing property, but not the value of what that property should be (IE clientID). 

Maybe this is giving a bit too much information as the person generating the token should be aware of all the required details for it to be considered valid